File: cerms-web/src/api/staff.ts
import { api } from '@/lib/axios'
import type { PaginatedList } from './services'

type ApiResponse<T> = {
  success: boolean
  data: T
  errors?: string[]
}

export type EmployeeCategory =
  | 'Operator'
  | 'OfficeStaff'
  | 'Manager'
  | 'Mechanic'
  | 'Helper'
  | 'Other'

export type EmploymentStatus = 'Active' | 'Inactive' | 'Suspended' | 'Resigned'

export const employeeCategoryValues = [
  'Operator',
  'OfficeStaff',
  'Manager',
  'Mechanic',
  'Helper',
  'Other',
] as const

export const employmentStatusValues = [
  'Active',
  'Inactive',
  'Suspended',
  'Resigned',
] as const

export type StaffDto = {
  id: string
  staffCode: string
  firstName: string
  lastName: string
  displayName: string
  employeeCategory: number
  mobileNo: string
  designation: string
  employmentStatus: number
  hasUserAccount: boolean
}

export type AssetClassDto = {
  id: string
  name: string
  description?: string | null
  isActive: boolean
}

export type StaffDetailDto = StaffDto & {
  gender: string
  dateOfBirth: string
  photoUrl?: string | null
  alternateMobileNo?: string | null
  email: string
  addressLine1: string
  addressLine2?: string | null
  city: string
  state: string
  pincode: string
  emergencyContactName: string
  emergencyContactNumber: string
  joiningDate: string
  relievingDate?: string | null
  department: string
  licenseNumber?: string | null
  licenseCategory?: string | null
  licenseExpiryDate?: string | null
  experienceYears?: number | null
  operatorGrade?: string | null
  dailyWage?: number | null
  salary?: number | null
  aadhaarNo?: string | null
  panNo?: string | null
  licenseDocumentUrl?: string | null
  idProofUrl?: string | null
  remarks?: string | null
  allowedAssetClasses: AssetClassDto[]
  linkedUserId?: string | null
}

export type StaffFormPayload = {
  firstName: string
  lastName: string
  displayName: string
  gender: string
  dateOfBirth: string
  mobileNo: string
  alternateMobileNo?: string | null
  email: string
  addressLine1: string
  addressLine2?: string | null
  city: string
  state: string
  pincode: string
  emergencyContactName: string
  emergencyContactNumber: string
  employeeCategory: number
  joiningDate: string
  relievingDate?: string | null
  employmentStatus: number
  designation: string
  department: string
  photoUrl?: string | null
  licenseNumber?: string | null
  licenseCategory?: string | null
  licenseExpiryDate?: string | null
  experienceYears?: number | null
  operatorGrade?: string | null
  dailyWage?: number | null
  salary?: number | null
  aadhaarNo?: string | null
  panNo?: string | null
  licenseDocumentUrl?: string | null
  idProofUrl?: string | null
  remarks?: string | null
  allowedAssetClassIds?: string[]
}

export type GetStaffParams = {
  pageNumber?: number
  pageSize?: number
  searchTerm?: string
  employeeCategory?: number
  employmentStatus?: number
}

export type StaffLookupDto = {
  id: string
  staffCode: string
  displayName: string
  employeeCategory: number
}

export const getStaffs = async (params?: GetStaffParams) => {
  const { data } = await api.get<ApiResponse<PaginatedList<StaffDto>>>('/staffs', { params })
  return data.data
}

export const getStaffById = async (id: string) => {
  const { data } = await api.get<ApiResponse<StaffDetailDto>>(`/staffs/${id}`)
  return data.data
}

export const createStaff = async (payload: StaffFormPayload) => {
  const { data } = await api.post<ApiResponse<StaffDetailDto>>('/staffs', payload)
  return data.data
}

export const updateStaff = async (id: string, payload: StaffFormPayload & { id: string }) => {
  const { data } = await api.put<ApiResponse<StaffDetailDto>>(`/staffs/${id}`, payload)
  return data.data
}

export const deactivateStaff = async (id: string) => {
  const { data } = await api.post<ApiResponse<unknown>>(`/staffs/${id}/deactivate`)
  return data.data
}

export const getAssetClasses = async () => {
  const { data } = await api.get<ApiResponse<AssetClassDto[]>>('/staffs/asset-classes')
  return data.data
}

export const getStaffWithoutUser = async () => {
  const { data } = await api.get<ApiResponse<StaffLookupDto[]>>('/staffs/without-user')
  return data.data
}

File: cerms-api/src/CERMS.Domain/Entities/Staff.cs
using CERMS.Domain.Common;
using CERMS.Domain.Enums;
using System;
using System.Collections.Generic;

namespace CERMS.Domain.Entities;

public class Staff : BaseEntity
{
    // Basic Details
    public string StaffCode { get; private set; }
    public string FirstName { get; private set; }
    public string LastName { get; private set; }
    public string DisplayName { get; private set; }
    public string Gender { get; private set; }
    public DateTime DateOfBirth { get; private set; }
    public string? PhotoUrl { get; private set; }

    // Contact Details
    public string MobileNo { get; private set; }
    public string? AlternateMobileNo { get; private set; }
    public string Email { get; private set; }

    // Address
    public string AddressLine1 { get; private set; }
    public string? AddressLine2 { get; private set; }
    public string City { get; private set; }
    public string State { get; private set; }
    public string Pincode { get; private set; }

    // Emergency Contact
    public string EmergencyContactName { get; private set; }
    public string EmergencyContactNumber { get; private set; }

    // Employment details
    public EmployeeCategory EmployeeCategory { get; private set; }
    public DateTime JoiningDate { get; private set; }
    public DateTime? RelievingDate { get; private set; }
    public EmploymentStatus EmploymentStatus { get; private set; }
    public string Designation { get; private set; }
    public string Department { get; private set; }

    // Operator Details (Only applicable if category == Operator)
    public string? LicenseNumber { get; private set; }
    public string? LicenseCategory { get; private set; }
    public DateTime? LicenseExpiryDate { get; private set; }
    public int? ExperienceYears { get; private set; }
    public string? OperatorGrade { get; private set; }

    // Financial & Identity (Optional)
    public decimal? DailyWage { get; private set; }
    public decimal? Salary { get; private set; }
    public string? AadhaarNo { get; private set; }
    public string? PANNo { get; private set; }

    // Documents (Urls/Paths)
    public string? LicenseDocumentUrl { get; private set; }
    public string? IdProofUrl { get; private set; }
    public string? Remarks { get; private set; }

    // Junction list
    public ICollection<StaffAssetClass> AllowedAssetClasses { get; private set; } = new List<StaffAssetClass>();
    
    // Linked User ID (One Staff = Max One User)
    public Guid? UserId { get; set; }

    protected Staff() { }

    public Staff(
        string staffCode, string firstName, string lastName, string displayName, string gender, DateTime dob,
        string mobileNo, string email, string addressLine1, string city, string state, string pincode,
        string emergencyContactName, string emergencyContactNumber, EmployeeCategory employeeCategory,
        DateTime joiningDate, string designation, string department, string? photoUrl = null)
    {
        StaffCode = staffCode;
        FirstName = firstName;
        LastName = lastName;
        DisplayName = displayName;
        Gender = gender;
        DateOfBirth = dob;
        MobileNo = mobileNo;
        Email = email;
        AddressLine1 = addressLine1;
        City = city;
        State = state;
        Pincode = pincode;
        EmergencyContactName = emergencyContactName;
        EmergencyContactNumber = emergencyContactNumber;
        EmployeeCategory = employeeCategory;
        JoiningDate = joiningDate;
        EmploymentStatus = EmploymentStatus.Active;
        Designation = designation;
        Department = department;
        PhotoUrl = photoUrl;
    }

    public void ConfigureOperatorDetails(string? licenseNumber, string? licenseCategory, DateTime? expiryDate, int? years, string? grade)
    {
        LicenseNumber = licenseNumber;
        LicenseCategory = licenseCategory;
        LicenseExpiryDate = expiryDate;
        ExperienceYears = years;
        OperatorGrade = grade;
    }

    public void UpdateFinancialsAndIdentity(decimal? dailyWage, decimal? salary, string? aadhaar, string? pan, string? remarks = null, string? licenseDoc = null, string? idProof = null)
    {
        DailyWage = dailyWage;
        Salary = salary;
        AadhaarNo = aadhaar;
        PANNo = pan;
        Remarks = remarks;
        LicenseDocumentUrl = licenseDoc;
        IdProofUrl = idProof;
    }

    public void UpdateDetails(
        string firstName, string lastName, string displayName, string gender, DateTime dob,
        string mobileNo, string? alternateMobile, string email, string addressLine1, string? addressLine2, string city, string state, string pincode,
        string emergencyContactName, string emergencyContactNumber, EmployeeCategory employeeCategory,
        DateTime joiningDate, DateTime? relievingDate, EmploymentStatus status, string designation, string department, string? photoUrl = null)
    {
        FirstName = firstName;
        LastName = lastName;
        DisplayName = displayName;
        Gender = gender;
        DateOfBirth = dob;
        MobileNo = mobileNo;
        AlternateMobileNo = alternateMobile;
        Email = email;
        AddressLine1 = addressLine1;
        AddressLine2 = addressLine2;
        City = city;
        State = state;
        Pincode = pincode;
        EmergencyContactName = emergencyContactName;
        EmergencyContactNumber = emergencyContactNumber;
        EmployeeCategory = employeeCategory;
        JoiningDate = joiningDate;
        RelievingDate = relievingDate;
        EmploymentStatus = status;
        Designation = designation;
        Department = department;
        PhotoUrl = photoUrl;
        Update();
    }

    public void Deactivate()
    {
        EmploymentStatus = EmploymentStatus.Inactive;
        Update();
    }

    public void Activate()
    {
        EmploymentStatus = EmploymentStatus.Active;
        Update();
    }
}
